crates/polars-stream/src/nodes/reduce.rs

```diff
@@ -0,0 +1,205 @@
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+use polars_core::frame::DataFrame;
+use polars_core::schema::SchemaRef;
+use polars_error::PolarsResult;
+use polars_expr::prelude::{ExecutionState, PhysicalExpr};
+use polars_expr::reduce::Reduction;
+
+use crate::async_executor::{JoinHandle, TaskPriority, TaskScope};
+use crate::async_primitives::pipe::{Receiver, Sender};
+use crate::graph::PortState;
+use crate::morsel::{Morsel, MorselSeq};
+use crate::nodes::ComputeNode;
+
+// All reductions in a single operation.
+// `select(sum, min) -> vec![sum, min]
+type ReduceSet = Vec<Box<dyn Reduction>>;
+
+enum ReduceState {
+    Sink {
+        inputs: Vec<Arc<dyn PhysicalExpr>>,
+        reductions: Arc<Mutex<Vec<Box<dyn Reduction>>>>,
+    },
+    Source(Mutex<Option<DataFrame>>),
+    Done,
+}
+
+pub struct ReduceNode {
+    // Reductions that are ready to finalize
+    full: Arc<Mutex<Vec<ReduceSet>>>,
+    state: ReduceState,
+    output_schema: SchemaRef,
+}
+
+impl ReduceNode {
+    pub fn new(
+        inputs: Vec<Arc<dyn PhysicalExpr>>,
+        reductions: ReduceSet,
+        output_schema: SchemaRef,
+    ) -> Self {
+        Self {
+            state: ReduceState::Sink {
+                inputs,
+                reductions: Arc::new(Mutex::new(reductions)),
+            },
+            output_schema,
+            full: Default::default(),
+        }
+    }
+
+    fn spawn_sink<'env, 's>(
+        &'env self,
+        scope: &'s TaskScope<'s, 'env>,
+        _pipeline: usize,
+        recv: &mut [Option<Receiver<Morsel>>],
+        _send: &mut [Option<Sender<Morsel>>],
+        state: &'s ExecutionState,
+    ) -> JoinHandle<PolarsResult<()>> {
```

Comment:
For consistency and early catching of bugs I always make sure to assert that `recv` and `send` have the expected lengths in the `spawn` functions, could you add that here?